File: c/Flyweight_Image.c
/// @file
/// @brief
/// Implementation of the Flyweight_ImageList and Flyweight_Image structures, the
/// latter which wraps the big resource handle and a Flyweight_Context as used
/// in the @ref flyweight_pattern.

#include <stdlib.h>

#include "Flyweight_Image.h"

///////////////////////////////////////////////////////////////////////////////
// Flyweight_ImageList_Clear()
///////////////////////////////////////////////////////////////////////////////
void Flyweight_ImageList_Clear(Flyweight_ImageList* imageList)
{
    if (imageList != NULL && imageList->images != NULL && imageList->images_count != 0)
    {
        free(imageList->images);
        imageList->images = NULL;
        imageList->images_count = 0;
    }
}


///////////////////////////////////////////////////////////////////////////////
// Flyweight_ImageList_Clear()
///////////////////////////////////////////////////////////////////////////////
void Flyweight_ImageList_Add(Flyweight_ImageList* imageList, Flyweight_Image* image)
{
    Flyweight_Image* new_list = NULL;

    if (imageList != NULL && image != NULL)
    {
        if (imageList->images == NULL)
        {
            new_list = malloc(sizeof(Flyweight_Image));
        }
        else
        {
            size_t newSize = (imageList->images_count + 1) * sizeof(Flyweight_Image);
            new_list = realloc(imageList->images, newSize);
        }

        if (new_list != NULL)
        {
            imageList->images = new_list;
            imageList->images[imageList->images_count] = *image;
            imageList->images_count++;
        }
    }
}

///// <summary>
///// Constructor (accessible only to the class factory).
///// </summary>
///// <param name="resource">The resource being represented by this class.  This
///// takes advantage of the fact that C# classes are always passed by reference so
///// no actual copying of the resource is done.</param>
///// <param name="context">A Flyweight_Context instance containing the context for this
///// Flyweight_Class instance.</param>
//Flyweight_Class(BigResource* resource, Flyweight_Context context);
//
///// <summary>
///// Retrieve the context for this class instance.
///// </summary>
//Flyweight_Context Context();
//
//
///// <summary>
///// Set the context for this class instance.
///// </summary>
//void SetContext(Flyweight_Context context);
//
//
///// <summary>
///// Retrieve the "image" width from underlying big resource.
///// </summary>
//int ImageWidth();
//
//
///// <summary>
///// Retrieve the "image" height from underlying big resource.
///// </summary>
//int ImageHeight();
//
///// <summary>
///// Render the image associated with this flyweight instance into the given
///// display at the given position.
///// </summary>
///// <param name="display">A list of character arrays representing the display.</param>
///// <param name="offset_x">Offset from left edge of big resource to the
///// left edge of the image to render.</param>
///// <param name="image_width">Width of image to render.</param>
///// <param name="image_height">Height of image to render.</param>
///// <param name="position_x">leftmost position within the display to place
///// the upper left corner of the image,</param>
///// <param name="position_y">topmost position within the display to place
///// the upper left corner of the image.</param>
///// <remarks>
///// Note that although this class has access to the context, it really
///// shouldn't so as to keep a loose coupling between the context and
///// this class.  Therefore, the entity calling Render() will pass the
///// position in as parameters even though that position is likely
///// coming from the context.
///// </remarks>
//void Render(std::vector<std::vector<char>>& display, int offset_x, int image_width, int image_height, int position_x, int position_y);

File: rust/design_pattern_examples_rust/src/flyweight/flyweight_image.rs
//! Contains the FlyweightImage struct that holds the ID of the big resource
//! "image" to draw from and a FlyWeightContext struct that specifies the
//! offsets to the Flyweight along with the position of the Flyweight "image"
//! within a "display".

//-----------------------------------------------------------------------------

use super::flyweight_context::FlyweightContext;
use super::flyweight_display::Display;
use super::flyweight_bigresource::BigResource;
use super::flyweight_bigresource_manager::BigResourceManager;

//-----------------------------------------------------------------------------

/// Represents an image that associates a context with a big resource.
pub struct FlyweightImage {
    /// The big resource being referenced by this flyweight image.  This is
    /// represented by a handle to the big resource.
    pub big_resource_id: usize,

    /// The context associated with this image.  The calling entity
    /// uses this context to manipulate the position of the image.
    pub context: FlyweightContext,
}

impl FlyweightImage {
    /// Render the image associated with this flyweight instance into the given
    /// display at the given position.
    ///
    /// # Parameters
    /// - resource_manager
    ///
    ///   The BigResourceManager from which to get the BigResource "image" to
    ///   use
    /// - display
    ///
    ///   The display to render to
    pub fn render(&self, resource_manager: &BigResourceManager, display: &mut Display) {
        let resource: &BigResource = match resource_manager.get_resource(self.big_resource_id) {
            Some(r) => r,
            None =>  {
                eprintln!("Unable to get big resource with id {}.  Cannot render images.", self.big_resource_id);
                return
            }
        };
        
        resource.render(display, self.context.offset_x_to_image,
        self.context.image_width, self.context.image_height,
        self.context.position_x as isize, self.context.position_y as isize);
    }
}
